### PR TITLE
bacpop-126 Error handling

### PR DESCRIPTION
**Pull request by EmmaLRussell**
This branch implements error handling in the back end. The approach is largely copied from WODIN, with some minor modifications. Key additions in this branch:

- `handleError` method which is registered with the app to handle any uncaught errors. This:
  - decides whether an error's message should be returned to front end - for unexpected errors, generates a unique id to display to user rather than the raw error. Error details from API are written to the response - we assume that JSON errors from beebop_py can be displayed to the user. (Although given that some API responses are coming back as HTML maybe this isn't currently true!)
  -  sets custom error values on the request, so that they are logged
  - returns standard porcelain-style error response
- `BeebopError` type for "expected" errors
- registers a custom logging format with `morgan` so that the custom error values are logged
- errors found in the course of processing requests asynchronously (e.g. API errors) should should be handled by invoking `handleError`, so that all errors will be handled (logged and written to response) generically by `handleError`. 

Unlike in WODIN, the backend is entirely JSON-based (all HTML is generated in the front end) so we do not need to distinguish between requests which expect HTML and JSON responses. 

I've added some integration tests for error responses - I've left adding integration tests for the logging part to another ticket, bacpop-127 to keep this changeset size down. 

The other change made in this branch are scripts to run up the deps, server and client outside pm2. 

Can be tested manually to an extent. For example
- run the dependencies, server and client (using the new scripts or pm2).
- Get an analysis ready to run by uploading a sample file
- Kill the beebop_py docker container
- Hit 'Start analysis' 
- The server should fail to connect to the API and should log the failure to connect in the server console. The response returned to the front end, observable in Network tab in the browser, should include unique error id instead of error details. That id should also have been logged in the server console. (You actually get two errors, one for the `poppunk` request and the second for the `status` request.)

### CODE REVIEW

File: app/server/src/utils.ts [hunk 1]
```diff
@@ -0,0 +1,27 @@
+import {BeebopError} from "./errors/beebopError";
+import {APIResponse} from "./types/responseTypes";
+
+export function throwAPIError(response, error: any) {
+    if (error.response) {
+        const apiResponse = error.response.data.error as APIResponse<any>;
```
Comment (EmmaLRussell) [inline app/server/src/utils.ts:6]: There seems to be an extra top level `error` on the response sent back by `beebop_py` for some reason - should probably fix this in another ticket, as it's a bit confusing!

File: app/server/src/errors/handleError.ts [hunk 2]
```diff
@@ -0,0 +1,31 @@
+import { Request, Response } from "express";
+import { uid } from "uid";
+import { reqWithError } from "../logging";
+import { BeebopError } from "./beebopError";
+
+function sendError(response, error, status = 500) {
```
Comment (EmmaLRussell) [inline app/server/src/errors/handleError.ts:6]: `handleError` should be the only place where we write error responses, so moved `sendError` to be internal to this module.

File: app/server/src/routes/routes.ts [hunk 3]
```diff
@@ -225,7 +227,7 @@ export const apiEndpoints = (config => ({
                 for (const sample of projectSamples) {
                     const apiSample = apiData.samples.find(s => s.hash === sample.hash);
                     if (!apiSample) {
-                        throw Error(`Sample with hash ${sample.hash} was not in API response`);
+                        throw new BeebopError(`Sample with hash ${sample.hash} was not in API response`, "Invalid data", 500, true);
```
Comment (EmmaLRussell) [inline app/server/src/routes/routes.ts:230]: Arguably this shouldn't return error message to front end as this isn't a typical 'Bad Request' but it's been useful to have an example of a BeebopError which has this flag set to true.

File: app/server/tsconfig.json [hunk 4]
```diff
@@ -1,5 +1,6 @@
 {
     "compilerOptions": {
+        "target": "es2019",
```
Comment (EmmaLRussell) [inline app/server/tsconfig.json:3]: Needed to add this to allow jest to run tests of code using null-coalescing operator

File: app/server/src/errors/handleError.ts [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
```
Comment (richfitz) [inline app/server/src/errors/handleError.ts:6]: Any need for a default here as it's always called with arg in place?

Comment (richfitz) [reply app/server/src/errors/handleError.ts:6]: Also, any reason why this function does not have type annotations?

File: app/server/tsconfig.json (hunk 4 above)
Comment (richfitz) [reply app/server/tsconfig.json:3]: we can presumably target whatever we want, no need to avoid new features

**Review by richfitz — APPROVED** [review]
Sorry, again not much to add here, but I see how this fits together at least (and obviously works elsewhere), and can see it working with the logs showing:

```
Config path: src/resources (/home/rfitzjoh/Documents/Projects/bacpop/beebop/app/server/src/resources)
listening on 4000
Connected to Redis server redis://localhost:6379
::1  GET /user 200 63 - 4.585 ms   
::1  GET /user 304  - 1.252 ms   
::1  GET /login/google 302 0 - 1.461 ms   
::1  GET /return/google?code=4%2F0AbUR2VN9tWN3pw-uz4T6UFArSteHwJka5aHFrPH-8-l5x9g9lYisq4tY6o5duFARBd4Miw&scope=profile+https%3A%2F%2Fwww.googleapis.com%2Fauth%2Fuserinfo.profile 302 86 - 150.393 ms   
::1  GET /user 200 104 - 0.845 ms   
::1  GET /user 304  - 0.614 ms   
::1  GET /projects 200 42 - 1.351 ms   
::1  GET /user 304  - 1.022 ms   
::1  OPTIONS /project 204 0 - 1.904 ms   
::1  POST /project 200 74 - 7.178 ms   
::1  GET /user 304  - 0.815 ms   
::1  OPTIONS /project/068b6ddaf4f6ec08c6777d3e0d867bdd/amr/f3d9b387e311d5ab59a8c08eb3545dbb 204 0 - 0.279 ms   
::1  POST /project/068b6ddaf4f6ec08c6777d3e0d867bdd/amr/f3d9b387e311d5ab59a8c08eb3545dbb 200 44 - 1.685 ms   
::1  OPTIONS /project/068b6ddaf4f6ec08c6777d3e0d867bdd/amr/e868c76fec83ee1f69a95bd27b8d5e76 204 0 - 0.401 ms   
::1  POST /project/068b6ddaf4f6ec08c6777d3e0d867bdd/amr/e868c76fec83ee1f69a95bd27b8d5e76 200 44 - 1.303 ms   
::1  OPTIONS /project/068b6ddaf4f6ec08c6777d3e0d867bdd/amr/346d9bf1dda11e87b248e72e7fd85df9 204 0 - 0.288 ms   
::1  POST /project/068b6ddaf4f6ec08c6777d3e0d867bdd/amr/346d9bf1dda11e87b248e72e7fd85df9 200 44 - 1.891 ms   
::1  OPTIONS /poppunk 204 0 - 0.315 ms   
::1  POST /poppunk 500 169 - 17.825 ms Unexpected error An unexpected error occurred. Please contact support and quote error code 68b6ddaf4f6 Error: Could not connect to API: Error: connect ECONNREFUSED 127.0.0.1:5000
    at handleAPIError (/home/rfitzjoh/Documents/Projects/bacpop/beebop/app/server/src/utils.ts:17:25)
    at /home/rfitzjoh/Documents/Projects/bacpop/beebop/app/server/src/routes/routes.ts:187:35
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async /home/rfitzjoh/Documents/Projects/bacpop/beebop/app/server/src/routes/routes.ts:175:13
    at async exports.default (/home/rfitzjoh/Documents/Projects/bacpop/beebop/app/server/src/errors/asyncHandler.ts:7:9)
    at async runPoppunk (/home/rfitzjoh/Documents/Projects/bacpop/beebop/app/server/src/routes/routes.ts:169:9)
::1  OPTIONS /status 204 0 - 0.346 ms   
::1  POST /status 500 169 - 2.220 ms Unexpected error An unexpected error occurred. Please contact support and quote error code 8b6ddaf4f6e Error: Could not connect to API: Error: connect ECONNREFUSED 127.0.0.1:5000
    at handleAPIError (/home/rfitzjoh/Documents/Projects/bacpop/beebop/app/server/src/utils.ts:17:25)
    at /home/rfitzjoh/Documents/Projects/bacpop/beebop/app/server/src/routes/routes.ts:249:31
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at async getStatus (/home/rfitzjoh/Documents/Projects/bacpop/beebop/app/server/src/routes/routes.ts:246:9)
```

with your example.

File: app/server/src/errors/handleError.ts (hunk 5 above)
Comment (EmmaLRussell) [reply app/server/src/errors/handleError.ts:6]: Went through a couple of iterations, I think there was a use of the default value at one point. Will remove, and add some type annotations too.
